presto-main/src/main/java/com/facebook/presto/sql/planner/EqualityInference.java

```diff
@@ -56,22 +55,17 @@
 public class EqualityInference
 {
     // Ordering used to determine Expression preference when determining canonicals
-    private static final Ordering<Expression> CANONICAL_ORDERING = Ordering.from(new Comparator<Expression>()
-    {
-        @Override
-        public int compare(Expression expression1, Expression expression2)
-        {
-            // Current cost heuristic:
-            // 1) Prefer fewer input symbols
-            // 2) Prefer smaller expression trees
-            // 3) Sort the expressions alphabetically - creates a stable consistent ordering (extremely useful for unit testing)
-            // TODO: be more precise in determining the cost of an expression
-            return ComparisonChain.start()
-                    .compare(SymbolsExtractor.extractAll(expression1).size(), SymbolsExtractor.extractAll(expression2).size())
-                    .compare(SubExpressionExtractor.extract(expression1).size(), SubExpressionExtractor.extract(expression2).size())
-                    .compare(expression1.toString(), expression2.toString())
-                    .result();
-        }
+    private static final Ordering<Expression> CANONICAL_ORDERING = Ordering.from((expression1, expression2) -> {
+        // Current cost heuristic:
+        // 1) Prefer fewer input symbols
+        // 2) Prefer smaller expression trees
+        // 3) Sort the expressions alphabetically - creates a stable consistent ordering (extremely useful for unit testing)
+        // TODO: be more precise in determining the cost of an expression
+        return ComparisonChain.start()
+                .compare(SymbolsExtractor.extractAll(expression1).size(), SymbolsExtractor.extractAll(expression2).size())
+                .compare(SubExpressionExtractor.extract(expression1).size(), SubExpressionExtractor.extract(expression2).size())
+                .compare(expression1.toString(), expression2.toString())
```

Comment:
Since you use `ComparisonChain`, this code is as effective as, but more complex than, equivalent using `Ordering.natural().onResultOf(expression -> comparisonKey)`.

I think it would be better not to call `SubExpressionExtractor.extract` if comparing `SymbolsExtractor.extractAll` was sufficient to determine the result.
(note: `ComparisonChain` is perfect when providing values for comparison is cheap)